samtranslator/model/api/api_generator.py

```diff
@@ -518,6 +532,134 @@ def _add_auth(self):
+        if not all(key in UsagePlanProperties._fields for key in usage_plan_properties.keys()):
+            raise InvalidResourceException(self.logical_id, "Invalid property for 'UsagePlan'")
+
+        create_usage_plan = usage_plan_properties.get("CreateUsagePlan")
+        usage_plan = None
+        api_key = None
+        usage_plan_key = None
+
+        if create_usage_plan is None:
+            raise InvalidResourceException(self.logical_id, "'CreateUsagePlan' is a required field for UsagePlan")
+        if create_usage_plan not in ["SHARED", "PER_API", "NONE"]:
+            raise InvalidResourceException(
+                self.logical_id, "'CreateUsagePlan' accepts only NONE, PER_API and SHARED values"
+            )
+
+        if create_usage_plan == "NONE":
+            return []
+
+        # create usage plan for this api only
+        elif usage_plan_properties.get("CreateUsagePlan") == "PER_API":
+            usage_plan_logical_id = self.logical_id + "UsagePlan"
+            usage_plan = ApiGatewayUsagePlan(logical_id=usage_plan_logical_id, depends_on=[self.logical_id])
+            api_stages = list()
+            api_stage = dict()
+            api_stage["ApiId"] = ref(self.logical_id)
+            api_stage["Stage"] = ref(rest_api_stage.logical_id)
+            api_stages.append(api_stage)
+            usage_plan.ApiStages = api_stages
+
+            api_key = self._construct_api_key(usage_plan_logical_id, create_usage_plan, rest_api_stage)
+            usage_plan_key = self._construct_usage_plan_key(usage_plan_logical_id, create_usage_plan, api_key)
+
+        # create a usage plan for all the Apis
+        elif create_usage_plan == "SHARED":
+            usage_plan_logical_id = "ServerlessUsagePlan"
+            ApiGenerator.depends_on_shared.append(self.logical_id)
+            usage_plan = ApiGatewayUsagePlan(
+                logical_id=usage_plan_logical_id, depends_on=ApiGenerator.depends_on_shared
+            )
+            api_stage = dict()
+            api_stage["ApiId"] = ref(self.logical_id)
+            api_stage["Stage"] = ref(rest_api_stage.logical_id)
+            ApiGenerator.api_stages_shared.append(api_stage)
+            usage_plan.ApiStages = ApiGenerator.api_stages_shared
+
+            api_key = self._construct_api_key(usage_plan_logical_id, create_usage_plan, rest_api_stage)
+            usage_plan_key = self._construct_usage_plan_key(usage_plan_logical_id, create_usage_plan, api_key)
+
+        if usage_plan_properties.get("UsagePlanName"):
+            usage_plan.UsagePlanName = usage_plan_properties.get("UsagePlanName")
+        if usage_plan_properties.get("Description"):
+            usage_plan.Description = usage_plan_properties.get("Description")
+        if usage_plan_properties.get("Quota"):
+            usage_plan.Quota = usage_plan_properties.get("Quota")
+        if usage_plan_properties.get("Tags"):
+            usage_plan.Tags = usage_plan_properties.get("Tags")
```

Comment:
We might want to add any tags added to the API to this Usage Plan as well, what do you think?

Reply:
a usage plan can contain multiple apis so i think its better not to propagate api tags to usage plan